File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/java/detectors/warehouselocalizer/WareHouseLocalizerHandler.java
package com.zebra.aisuite_quickstart.java.detectors.warehouselocalizer;

import android.content.Context;
import android.util.Log;

import androidx.camera.core.ImageAnalysis;
import androidx.core.content.ContextCompat;

import com.zebra.ai.vision.detector.InferencerOptions;
import com.zebra.ai.vision.detector.Localizer;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class WareHouseLocalizerHandler {
    private static final String TAG = "WareHouseLocalizerHandler";
    private Localizer wareHouseLocalizer;       // For live preview
    private Localizer captureLocalizer;
    private final ExecutorService executor;
    private final ExecutorService captureExecutor = Executors.newSingleThreadExecutor();
    private final Context context;
    private WareHouseAnalyzer wareHouseAnalyzer;
    private final WareHouseAnalyzer.DetectionCallback callback;
    private final ImageAnalysis imageAnalysis;
    private final String mavenModelName = "warehouse-localizer";
    private final ModelLoadingCallback loadingCallback;

    // Model input sizes
    private static final int LIVE_PREVIEW_SIZE = 640;
    private static final int CAPTURE_SIZE = 1280; // Higher resolution for capture

    /**
     * Callback interface for model loading completion
     */
    public interface ModelLoadingCallback {
        void onLoadingComplete(boolean success);
    }

    public WareHouseLocalizerHandler(Context context, WareHouseAnalyzer.DetectionCallback callback, ImageAnalysis imageAnalysis, ModelLoadingCallback loadingCallback) {
        this.context = context;
        this.callback = callback;
        this.executor = Executors.newSingleThreadExecutor();
        this.imageAnalysis = imageAnalysis;
        this.loadingCallback = loadingCallback;

        initializeWareHouseLocalizer();
        initializeCaptureLocalizer();
    }

    public void initializeWareHouseLocalizer() {
        try {
            // Initialize live preview localizer with smaller input size
            Localizer.Settings liveLocalizerSettings = createLocalizerSettings(LIVE_PREVIEW_SIZE);
            createWareHouseLocalizerWithFallback(liveLocalizerSettings);
        } catch (Exception ex) {
            if (loadingCallback != null) {
                loadingCallback.onLoadingComplete(false);
            }
            Log.e(TAG, "Model Loading: WareHouse localizer returned with exception " + ex.getMessage());
        }
    }

    /**
     * Initializes the capture localizer with higher resolution settings
     */
    public void initializeCaptureLocalizer() {
        try {
            Localizer.Settings captureLocalizerSettings = createLocalizerSettings(CAPTURE_SIZE);
            createCaptureLocalizerWithFallback(captureLocalizerSettings);
        } catch (Exception ex) {
            if (loadingCallback != null) {
                loadingCallback.onLoadingComplete(false);
            }
            Log.e(TAG, "Capture localizer initialization failed: " + ex.getMessage());
        }
    }

    /**
     * Creates localizer settings with specified input size
     */
    private Localizer.Settings createLocalizerSettings(int inputSize) {
        Localizer.Settings localizerSettings = new Localizer.Settings(mavenModelName);
        Integer[] rpo = new Integer[3];
        rpo[0] = InferencerOptions.DSP;
        rpo[1] = InferencerOptions.CPU;
        rpo[2] = InferencerOptions.GPU;

        localizerSettings.inferencerOptions.runtimeProcessorOrder = rpo;
        localizerSettings.inferencerOptions.defaultDims.height = inputSize;
        localizerSettings.inferencerOptions.defaultDims.width = inputSize;

        return localizerSettings;
    }

    private void createWareHouseLocalizerWithFallback(Localizer.Settings localizerSettings) {
        long m_Start = System.currentTimeMillis();
        Localizer.getLocalizer(localizerSettings, executor).thenAccept(localizerInstance -> {
            wareHouseLocalizer = localizerInstance;
            if(captureLocalizer!=null) {
                if (loadingCallback != null) {
                    loadingCallback.onLoadingComplete(true);
                }
                attachAnalysisAfterModelLoading();
            }
            Log.d(TAG, "WareHouseLocalizer() obj creation time =" + (System.currentTimeMillis() - m_Start) + " milli sec and input size: " + localizerSettings.inferencerOptions.defaultDims.width);
        }).exceptionally(e -> {
            if (loadingCallback != null) {
                loadingCallback.onLoadingComplete(false);
            }
            Log.e(TAG, "Fatal error: localizer creation failed - " + e.getMessage());
            return null;
        });
    }

    private void createCaptureLocalizerWithFallback(Localizer.Settings localizerSettings) {
        long m_Start = System.currentTimeMillis();
        Localizer.getLocalizer(localizerSettings, captureExecutor).thenAccept(localizerInstance -> {
            captureLocalizer = localizerInstance;
            if(wareHouseLocalizer!=null) {
                if (loadingCallback != null) {
                    loadingCallback.onLoadingComplete(true);
                }
                attachAnalysisAfterModelLoading();
            }
            Log.d(TAG, "Capture WareHouseLocalizer created in " + (System.currentTimeMillis() - m_Start) + " ms");
        }).exceptionally(e -> {
            if (loadingCallback != null) {
                loadingCallback.onLoadingComplete(false);
            }
            Log.e(TAG, "Capture localizer creation failed: " + e.getMessage());
            return null;
        });
    }

    /**
     * Stops the executor service and disposes of both WareHouseLocalizers
     */
    public void stop() {
        executor.shutdownNow();
        captureExecutor.shutdownNow();
        if (wareHouseLocalizer != null) {
            wareHouseLocalizer.dispose();
            Log.d(TAG, "Live preview warehouse localizer disposed");
            wareHouseLocalizer = null;
        }
        if (captureLocalizer != null) {
            captureLocalizer.dispose();
            Log.d(TAG, "Capture warehouse localizer disposed");
            captureLocalizer = null;
        }
    }

    public WareHouseAnalyzer getWareHouseAnalyzer() {
        return wareHouseAnalyzer;
    }

    public Localizer getCaptureLocalizer() {
        return captureLocalizer;
    }

    public void attachAnalysisAfterModelLoading(){
        wareHouseAnalyzer = new WareHouseAnalyzer(callback, wareHouseLocalizer);
        imageAnalysis.setAnalyzer(ContextCompat.getMainExecutor(context), wareHouseAnalyzer);
    }
}

File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/kotlin/detectors/warehouselocalizer/WareHouseLocalizerHandler.kt
package com.zebra.aisuite_quickstart.kotlin.detectors.warehouselocalizer

import android.content.Context
import android.util.Log
import androidx.camera.core.ImageAnalysis
import androidx.core.content.ContextCompat
import com.zebra.ai.vision.detector.InferencerOptions
import com.zebra.ai.vision.detector.Localizer
import com.zebra.aisuite_quickstart.kotlin.CameraXLivePreviewActivity
import com.zebra.aisuite_quickstart.utils.CommonUtils
import java.util.concurrent.Executors

class WareHouseLocalizerHandler(
    private val context: Context,
    private val callback: CameraXLivePreviewActivity,
    private val imageAnalysis: ImageAnalysis,
    private val loadingCallback: ((Boolean) -> Unit)? = null
) {
    companion object {
        private const val TAG = "WareHouseLocalizerHandler"
        private const val CAPTURE_SIZE = 1280 // Higher resolution for capture
    }

    private var wareHouseLocalizer: Localizer? = null // For live preview
    var captureLocalizer: Localizer? = null // For capture mode
    private val executor = Executors.newSingleThreadExecutor()
    private val captureExecutor = Executors.newSingleThreadExecutor()
    var wareHouseAnalyzer: WareHouseAnalyzer? = null
    private val mavenModelName = "pallet-and-box-localizer"
    private val sharedPreferences = context.getSharedPreferences(CommonUtils.SETTINGS_PREFS, Context.MODE_PRIVATE)

    init {
        initializeWareHouseLocalizer()
        initializeCaptureLocalizer()
    }

    /**
     * Creates localizer settings with specified input size.
     *
     * @param inputSize The input dimension size for the localizer model.
     * @return Configured Localizer.Settings instance.
     */
    private fun createLocalizerSettings(inputSize: Int): Localizer.Settings {
        return Localizer.Settings(mavenModelName).apply {
            val rpo = arrayOf(
                InferencerOptions.DSP,
                InferencerOptions.CPU,
                InferencerOptions.GPU
            )
            inferencerOptions.runtimeProcessorOrder = rpo
            inferencerOptions.defaultDims.height = inputSize
            inferencerOptions.defaultDims.width = inputSize
        }
    }

    /**
     * Initializes the live preview WareHouse localizer with smaller input size for real-time processing.
     */
    private fun initializeWareHouseLocalizer() {
        val modelInputSize = sharedPreferences.getInt(CommonUtils.PREF_MODEL_INPUT_SIZE, 640)
        Log.d(TAG, " LivePreview Model Input Size: $modelInputSize")
        try {
            val liveLocalizerSettings = createLocalizerSettings(modelInputSize)
            createWareHouseLocalizer(liveLocalizerSettings)
        } catch (ex: Exception) {
            loadingCallback?.invoke(false)
            Log.e(
                TAG,
                "Model Loading: Pallet and Box Localizer returned with exception ${ex.message}"
            )
        }
    }

    /**
     * Initializes the capture localizer with higher resolution settings.
     */
    fun initializeCaptureLocalizer() {
        try {
            val captureLocalizerSettings = createLocalizerSettings(CAPTURE_SIZE)
            createCaptureLocalizer(captureLocalizerSettings)
        } catch (ex: Exception) {
            loadingCallback?.invoke(false)
            Log.e(TAG, "Capture localizer initialization failed: ${ex.message}")
        }
    }

    /**
     * Creates the live preview WareHouse Localizer instance.
     * Only notifies loading complete and attaches analyzer when both models are loaded.
     */
    private fun createWareHouseLocalizer(localizerSettings: Localizer.Settings) {
        val startTime = System.currentTimeMillis()
        Localizer.getLocalizer(localizerSettings, executor)
            .thenAccept { localizerInstance ->
                wareHouseLocalizer = localizerInstance

                if (captureLocalizer != null) {
                    loadingCallback?.invoke(true)
                    attachAnalysisAfterModelLoading()
                }

                Log.d(
                    TAG,
                    "Pallet and Box Localizer() obj creation time = ${System.currentTimeMillis() - startTime} ms" +
                            " and input size: ${localizerSettings.inferencerOptions.defaultDims.width}"
                )
            }
            .exceptionally { e ->
                loadingCallback?.invoke(false)
                Log.e(TAG, "Fatal error: Pallet and Box Localizer creation failed - ${e.message}")
                null
            }
    }

    /**
     * Creates the capture WareHouse Localizer instance.
     * Only notifies loading complete and attaches analyzer when both models are loaded.
     */
    private fun createCaptureLocalizer(localizerSettings: Localizer.Settings) {
        val startTime = System.currentTimeMillis()
        Localizer.getLocalizer(localizerSettings, captureExecutor)
            .thenAccept { localizerInstance ->
                captureLocalizer = localizerInstance

                if (wareHouseLocalizer != null) {
                    loadingCallback?.invoke(true)
                    attachAnalysisAfterModelLoading()
                }

                Log.d(
                    TAG,
                    "Capture Pallet and Box Localizer created in ${System.currentTimeMillis() - startTime} ms"
                )
            }
            .exceptionally { e ->
                loadingCallback?.invoke(false)
                Log.e(TAG, "Capture Pallet and Box Localizer creation failed: ${e.message}")
                null
            }
    }

    /**
     * Attaches the WareHouseAnalyzer to the ImageAnalysis once both models are loaded.
     */
    private fun attachAnalysisAfterModelLoading() {
        wareHouseAnalyzer = WareHouseAnalyzer(callback, wareHouseLocalizer)
        imageAnalysis.setAnalyzer(ContextCompat.getMainExecutor(context), wareHouseAnalyzer!!)
    }

    /**
     * Stops the executor services and disposes of both WareHouse Localizer instances,
     * releasing any resources held.
     */
    fun stop() {
        executor.shutdownNow()
        captureExecutor.shutdownNow()
        wareHouseLocalizer?.let {
            it.dispose()
            Log.d(TAG, "Live preview Pallet and Box Localizer disposed")
            wareHouseLocalizer = null
        }
        captureLocalizer?.let {
            it.dispose()
            Log.d(TAG, "Capture Pallet and Box Localizer disposed")
            captureLocalizer = null
        }
    }

}